### PR TITLE
Use gnu date if available

### PR DESCRIPTION
**Pull request by micahhausler**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Uses gnu `date` if available.  On macOS,  BSD `date` has different flags. When installing `coreutils` on macOS with homebrew, commands are [prefixed with a "g"](https://github.com/Homebrew/homebrew-core/blob/72ca636/Formula/coreutils.rb#L46).  This change is equivalent to how Kubernetes uses gnu `tar` in [`build/common.sh`](https://github.com/kubernetes/kubernetes/blob/e4e9c31218becac11f176cc824f5cc6b7a8036ac/build/common.sh#L222-L241)

#### Which issue(s) this PR fixes:

N/A

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
```

### CODE REVIEW

File: hack/lib/version.sh [hunk 1]
```diff
@@ -161,8 +161,14 @@ kube::version::ldflags() {
...
   }
+  DATE=date
+  if which gdate &>/dev/null; then
+      DATE=gdate
+  elif which gnudate &>/dev/null; then
+      DATE=gnudate
+  fi
 
-  add_ldflag "buildDate" "$(date ${SOURCE_DATE_EPOCH:+"--date=@${SOURCE_DATE_EPOCH}"} -u +'%Y-%m-%dT%H:%M:%SZ')"
+  add_ldflag "buildDate" "$(${DATE} ${SOURCE_DATE_EPOCH:+"--date=@${SOURCE_DATE_EPOCH}"} -u +'%Y-%m-%dT%H:%M:%SZ')"
```
Comment (BenTheElder) [inline hack/lib/version.sh:171]: Unlike `TAR` and `DOCKER` I think `DATE` is more likely to be confused with say, the actual date.
Maybe `GNU_DATE`?

**Comment by BenTheElder** [comment]
+1 in concept, but I think if we're depending on non-portable date flags we should fail with the wrong binary and check for it early like gnu tar.

**Comment by micahhausler** [comment]
> +1 in concept, but I think if we're depending on non-portable date flags we should fail with the wrong binary and check for it early like gnu tar.

Cool, I'll add a function similar to the tar check

**Comment by micahhausler** [comment]
There are several other `date` calls, but they're all compatible for both BSD and GNU date. Is it worth changing them? (This list might not be complete)
```
$ grep -E '(?:date\ (?:-u )?"?\+)|(?:\$\(date)' -r ./{hack,build}
./hack/cherry_pick_pull.sh:NEWBRANCHUNIQ="${NEWBRANCH}-$(date +%s)"
./hack/make-rules/test-e2e-kubeadm.sh:artifacts=${ARTIFACTS:-"/tmp/_artifacts/$(date +%y%m%dT%H%M%S)"}
./hack/make-rules/verify.sh:    start=$(date +%s)
./hack/make-rules/verify.sh:    local elapsed=$(($(date +%s) - start))
./hack/make-rules/test-e2e-node.sh:artifacts="${ARTIFACTS:-"/tmp/_artifacts/$(date +%y%m%dT%H%M%S)"}"
./hack/grab-profiles.sh:timestamp=$(date +%Y%m%d%H%M%S)
./hack/local-up-cluster.sh:  print_color "$1" "W$(date "+%m%d %H:%M:%S")]" 1
./hack/lib/etcd.sh:    ETCD_LOGFILE="${ARTIFACTS}/etcd.$(uname -n).$(id -un).log.DEBUG.$(date +%Y%m%d-%H%M%S).$$"
./hack/lib/logging.sh:  timestamp=$(date +"[%m%d %H:%M:%S]")
./hack/lib/logging.sh:  timestamp=$(date +"[%m%d %H:%M:%S]")
./hack/lib/swagger.sh:  TMPFILE="${TMPDIR:-/tmp}/types_swagger_doc_generated.$(date +%s).go"
./hack/lib/util.sh:  date "+%Y%m%d-%H%M%S"
./hack/print-workspace-status.sh:buildDate $(date \
./hack/jenkins/benchmark-dockerized.sh:   >(go run test/integration/benchmark/jsonify/main.go "${ARTIFACTS}/BenchmarkResults_benchmark_$(date -u +%Y-%m-%dT%H:%M:%SZ).json" || cat > /dev/null)
./build/root/Makefile:#    Defaults to "/tmp/_artifacts/$$(date +%y%m%dT%H%M%S)".
```

**Comment by BenTheElder** [comment]
We should probably just leave the ones that don't depend on this.

One thing that isn't clear though: do we depend on it here? Or only if SOURCE_DATE_EPOCH was set? If the latter it probably shouldn't be prereq for building.

Ideally we also minimize the gnuisms we depend on, but we have precedent for things without excellent alternatives.

**Comment by micahhausler** [comment]
> We should probably just leave the ones that don't depend on this.

SGTM

> One thing that isn't clear though: do we depend on it here? Or only if SOURCE_DATE_EPOCH was set? If the latter it probably shouldn't be prereq for building.
> 
> Ideally we also minimize the gnuisms we depend on, but we have precedent for things without excellent alternatives.

AFAICT I think the only time it is required is when `SOURCE_DATE_EPOCH` is set and needs to be parsed. 

I'm having a bit of a hard time telling which tests eventually call `kube::version::ldflags`, is the only way to determine just trial and error by setting `import /build/common.sh`/`kube::build::ensure_gnu_date` in each bash script invoked by make targets?

**Comment by micahhausler** [comment]
@BenTheElder Putting a `kube::build::ensure_gnu_date()` in `build/common.sh` next to `kube::build::ensure_tar()`, results in a dependency loop.

`build/common.sh` sources `hack/lib/init.sh`  which sources `hack/lib/version.sh`, where `ensure_gnu_date` needs to be called. 

> With GNU tar (which we definitely need versus bsd tar), we ensure that we have the correct tool and fail if not.
We also check for this early in the build process instead of just-in-time so we can fail early if it's missing.

Would it make sense to define and then call `ensure_gnu_date` at the beginning of `hack/lib/version.sh`? It would then be called early, and would avoid the dependency loop.

File: hack/lib/version.sh [hunk 2]
```diff
@@ -27,6 +27,26 @@
 #    KUBE_GIT_MAJOR - The major part of the version
 #    KUBE_GIT_MINOR - The minor component of the version
 
+# Ensure GNU tar is present.
```
Comment (BenTheElder) [inline hack/lib/version.sh:30]: stale comment

File: hack/lib/version.sh [hunk 3]
```diff
@@ -27,6 +27,26 @@
...
+  GNU_DATE="date"
+  if which gdate &>/dev/null; then
+      GNU_DATE="gdate"
+  elif which gnudate &>/dev/null; then
+      GNU_DATE="gnudate"
+  fi
+  if ! "${GNU_DATE}" --version | grep -q GNU; then
+    echo "  !!! Cannot find GNU tar. Build on Linux or install GNU tar"
+    echo "      on Mac OS X (brew install coreutils)."
+    return 1
```
Comment (BenTheElder) [inline hack/lib/version.sh:44]: still not sure if this should be a hard failure or just a warning, it's only necessary if SOURCE_DATE_EPOCH so maybe only fail if SOURCE_DATE_EPOCH ?

**Review by BenTheElder — COMMENTED** [review]
sorry this dropped off my radar :(

**Comment by dims** [comment]
@micahhausler wanna update this? (there are 2 comments)

**Comment by BenTheElder** [comment]
```
Status: Downloaded newer image for koalaman/shellcheck-alpine@sha256:d6147f30864ddb7c9cf983fc277d345bc315e798e309ddf70062b194843ee252
In ./hack/lib/version.sh line 41:
    if [ -n ${SOURCE_DATE_EPOCH} ]; then
            ^------------------^ SC2070: -n doesn't work with unquoted arguments. Quote or use [[ ]].
            ^------------------^ SC2086: Double quote to prevent globbing and word splitting.
Did you mean: 
    if [ -n "${SOURCE_DATE_EPOCH}" ]; then
For more information:
  https://www.shellcheck.net/wiki/SC2070 -- -n doesn't work with unquoted arg...
  https://www.shellcheck.net/wiki/SC2086 -- Double quote to prevent globbing ...
```

**Pull request closed without merging**